src/data/padics/padic_norm.lean

```diff
@@ -298,6 +316,225 @@ theorem min_le_padic_val_rat_add {q r : ℚ}
+begin
+  by_cases a_split : (a = 0),
+  { subst a_split, simp [dvd2] },
+  { rcases dvd2 with ⟨k, rfl⟩,
+    rw nat.mul_div_cancel_left k (nat.pos_of_ne_zero a_split),
+    by_cases c_split : (c = 0),
+    { subst c_split, simp only [zero_dvd_iff] at dvd, subst dvd, trivial, },
+    { rcases dvd with ⟨k2, rfl⟩,
+      rw nat.mul_div_cancel_left k2 (nat.pos_of_ne_zero c_split),
+      rw nat.mul_comm (c * k2) k,
+      rw ← nat.mul_assoc k c k2,
+      have k2_nonzero : k2 ≠ 0,
+        { intros k2_zero,
+          subst k2_zero,
+          simp at a_split,
+          trivial, },
+      rw nat.mul_div_cancel _ (nat.pos_of_ne_zero k2_nonzero),
+      rw nat.mul_div_cancel _ (nat.pos_of_ne_zero c_split), }, },
+end
+
+lemma padic_val_nat_div' {p : ℕ} [p_prime : fact p.prime] :
+  ∀ {m : ℕ} (cpm : coprime p m) {b : ℕ} (dvd : m ∣ b), padic_val_nat p (b / m) = padic_val_nat p b
+| 0 := λ cpm b dvd, by
+  begin
+    exfalso,
+    have p_one : p = 1, by simpa using cpm,
+    simpa [p_one, nat.not_prime_one] using p_prime,
+  end
+| 1 := λ cpm b dvd, by simp
+| (n + 2) :=
+  let m := n + 2 in
+  let min_fac := nat.min_fac m in
+  let wf : m / (nat.min_fac m) < m := nat.div_lt_self (by linarith) (nat.prime.one_lt (nat.min_fac_prime (by linarith))) in
+  λ cpm b dvd,
+  begin
+    rw ← div_div_div (nat.min_fac_dvd m) dvd,
+    haveI: fact (nat.prime min_fac) := nat.min_fac_prime (by linarith),
+    have p_different : p ≠ min_fac,
+      { intros p_eq,
+        have bad : m.min_fac ∣ m := min_fac_dvd m,
+        rw p_eq at cpm,
+        have min_fac_big : 1 < min_fac := prime.one_lt (nat.min_fac_prime (by linarith)),
+        exact not_coprime_of_dvd_of_dvd min_fac_big (dvd_refl min_fac) bad cpm, },
+    have u : min_fac ∣ b / (m / min_fac),
+      { rcases dvd with ⟨k, rfl⟩,
+        rcases nat.min_fac_dvd m with ⟨l, pr⟩,
+        rw [pr, nat.mul_div_cancel_left _ (nat.min_fac_pos m)],
+        have pr' : (n + 2) = m.min_fac * l := pr,
+        rw [pr', mul_comm min_fac, mul_assoc],
+        have l_pos : 0 < l,
+          { cases lt_or_le 0 l,
+            { exact h },
+            { rw le_zero_iff_eq at h,
+              subst h,
+              simp only [add_eq_zero_iff, mul_zero] at pr',
+              trivial, }, },
+        rw nat.mul_div_cancel_left _ l_pos,
+        exact dvd.intro k rfl, },
+    rw padic_val_nat_div'' p_different u,
+    have w : (m / min_fac) ∣ b,
+      { rcases dvd with ⟨k, rfl⟩,
+        by_cases k_split : (k = 0),
+        { subst k_split, simp, },
+        { exact dvd_mul_of_dvd_left (div_dvd_of_dvd (nat.min_fac_dvd m)) _ }, },
+    rw padic_val_nat_div' (coprime.coprime_div_right cpm (min_fac_dvd m)) w,
+  end
+
+open_locale big_operators
+
+lemma prod_padic_val_nat_one (m : nat) :
+  ∏ p in finset.filter nat.prime (finset.range m), p ^ padic_val_nat p 1 = 1 :=
+by simp [padic_val_nat_of_one]
+
+lemma prod_pow_prime_padic_val_nat : ∀ (n : nat) (s : n ≠ 0) (m : nat) (pr : n < m),
```

Comment:
I think this should move to the Bertrand file. It's quite specialised, I think.
But I don't have a very strong opinion on this.

Reply:
Honestly I think of it as rather a defining property of `padic_val_nat`; I'm moderately confident it should be here.

Reply:
Ok, then let's leave it here.